Implement `all_linters()` wrapper to access all available linters

IndrajeetPatil opened the pull request:
Closes #1843

AshesITR commented:
Can we add integration with `use_lintr()`?

AshesITR commented:
We need much more thorough cross referencing and update our examples that use `linters_with_tags(NULL)` to use the new function.

A new user should be able to easily discover all three variants and understand how and when to use what.

IndrajeetPatil commented:
> We need much more thorough cross referencing and update our examples that use `linters_with_tags(NULL)` to use the new function.
> 
> A new user should be able to easily discover all three variants and understand how and when to use what.

Good idea.

I have 

- cross-referenced them heavily in the Roxygen docs via seealso tag
- changed the relevant code in README
- removed one example from `linters_with_tags()` docs as well
- changed relevant examples in lintr vignette

IndrajeetPatil commented on R/with.R:
```diff
@@ -76,10 +83,6 @@ modify_defaults <- function(defaults, ...) {
 #' linters <- linters_with_tags(tags = "package_development")
 #' names(linters)
 #'
-#' # Get all linters provided by lintr
-#' linters <- linters_with_tags(tags = NULL)
-#' names(linters)
```
Since our preferred way to do this now is via `all_linters()`.

AshesITR commented on R/with.R:
```diff
@@ -8,7 +8,12 @@
 #'   transformations are applied to the names: (1) replace `_` with `0` and (2) convert [tolower()].
-#' @seealso [linters_with_tags], [linters_with_defaults] for creating linter lists.
+#' @seealso
+#' - [linters_with_defaults] for basing off lintr's set of default linters.
```
linters_with_tags was removed here.

IndrajeetPatil replied:
Added it back.

Also added `@family` tag to make sure all related functions here are interlinked.

AshesITR commented on R/with.R:
```diff
@@ -117,6 +120,23 @@ linters_with_tags <- function(tags, ..., packages = "lintr", exclude_tags = "dep
+#' - [available_linters] to get a data frame of available linters.
+#' - [linters] for a complete list of linters available in lintr.
+#' @export
+all_linters <- function(packages = "lintr") {
```
This needs ellipsis to allow for configuration.
For example compare
`linters_with_tags(NULL, line_length_linter = NULL)`.

IndrajeetPatil replied:
Done.

AshesITR commented on README.md:
```diff
@@ -3,11 +3,11 @@
 ![](man/figures/demo.gif "lintr demo")
 
-`lintr` is complementary to [the `styler` package](https://github.com/r-lib/styler) which automatically restyles code, eliminating some of the problems that `lintr` can detect.
+`{lintr}` is complementary to [the `styler` package](https://github.com/r-lib/styler) which automatically restyles code, eliminating some of the problems that `{lintr}` can detect.
```
`styler` is also a package.

AshesITR commented on README.md:
````diff
@@ -26,26 +26,44 @@ remotes::install_github("r-lib/lintr")
+Load the library:
+
 ```R
+library(lintr)
````
I think we should encourage keeping `search()` clean, i.e. stick with 
 `lintr::lint_dir()` etc.

IndrajeetPatil replied:
Removed all library calls.

AshesITR commented on README.md:
````diff
@@ -26,26 +26,44 @@ remotes::install_github("r-lib/lintr")
+You can use helper functions from `{usethis}` package to generate GitHub action workflow files:
+
+```R
+library(usethis)
````
ditto.

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -199,14 +199,14 @@ knitr::kable(defaults_table)
 If you want to use all available linters, you can include this in your `.lintr` file:
 
 ```r
 linters: linters_with_defaults(
````
We should probably advertise using `all_linters(...)` and `linters_with_tags(...)` directly.

AshesITR commented on R/with.R:
```diff
@@ -117,6 +126,25 @@ linters_with_tags <- function(tags, ..., packages = "lintr", exclude_tags = "dep
+#' - [linters] for a complete list of linters available in lintr.
+#' @export
+all_linters <- function(packages = "lintr", ...) {
+  linters_with_tags(tags = NULL, packages = packages, ...)
```
What happens in the unlikely case that `tags` is given as a named argument?

IndrajeetPatil replied:
It will produce an error: 

``` r
library(lintr)
all_linters(tags = "default")
#> Error in linters_with_tags(tags = NULL, packages = packages, ...): formal argument "tags" matched by multiple actual arguments
```

<sup>Created on 2022-12-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

But I doubt anyone will do this because this assumes that the user has looked at the implementation details for this newly introduced function and figured out that `all_linters()` is a wrapper around `linters_with_tags()`, and is deliberately using this argument. But, in that case, they shouldn't be surprised by the error message they get.

AshesITR replied:
`modify_defaults(..., defaults = linters_with_tags(tags = NULL, packages = packages)` should not suffer from this bug.

AshesITR commented on README.md:
```diff
@@ -1,13 +1,14 @@
 # lintr
-[![R build status](https://github.com/r-lib/lintr/workflows/R-CMD-check/badge.svg)](https://github.com/r-lib/lintr/actions)
+
+[![R build status](https://github.com/r-lib/lintr/workflows/R-CMD-check/badge.svg)](https://github.com/r-lib/lintr/actions/)
```
Why is that?

IndrajeetPatil replied:
All Markdown elements should be separated by blank lines.

cf. https://yihui.org/en/2021/06/markdown-breath/

AshesITR replied:
And the trailing slash in the URL?

IndrajeetPatil replied:
Force of habit. I have reverted these changes. Although they won't cause any issues, the trailing slashes are unnecessary here.

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -216,7 +216,18 @@ If you want to use all available linters *except* a few, you can exclude them us
+Alternatively, you can also do:
+
+```r
+linters: linters_with_defaults(
````
What I had in mind was

```r
linters: all_linters(
    commented_code_linter = NULL,
    implicit_integer_linter = NULL
  )
```

IndrajeetPatil replied:
Ah, I see. 

I've changed the example.

AshesITR commented on README.md:
````diff
@@ -45,7 +46,17 @@ To see a list of linters included for each configuration:
+
+```R
+# in a project:
+usethis::use_github_action("lint-project")
````
Why split this off?

IndrajeetPatil replied:
Separation of concerns. 

Not every lintr user will concern themselves with integrating the workflow with GitHub actions (many might not even be using GitHub), so we should separate the documentation of how to set up config file from how to use lintr with GHA.

AshesITR replied:
Sounds fair, should we add a subsection then?
`### Setting up GitHub Actions` maybe?

AshesITR commented on man/all_linters.Rd:
```diff
@@ -0,0 +1,35 @@
+}
+\seealso{
+\itemize{
+\item \link{linters_with_defaults} for basing off lintr's set of default linters.
+\item \link{linters_with_tags} for basing off tags attached to linters, possibly across multiple packages.
+\item \link{available_linters} to get a data frame of available linters.
+\item \link{linters} for a complete list of linters available in lintr.
+}
+
+Other linter-config: 
+\code{\link{linters_with_tags}()},
+\code{\link{modify_defaults}()}
```
This looks quite odd (duplicated links)

IndrajeetPatil replied:
Removed the family tag. Redundant when the seealso tag is also used.

AshesITR commented on vignettes/lintr.Rmd:
```diff
@@ -216,7 +216,16 @@ If you want to use all available linters *except* a few, you can exclude them us
     commented_code_linter = NULL,
     implicit_integer_linter = NULL,
-    defaults = linters_with_tags(tags = NULL)
+    defaults = all_linters()
```
I wouldn't show this way to write the config. It's unnecessarily verbose.

IndrajeetPatil replied:
True. Removed it.

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -199,24 +199,23 @@ knitr::kable(defaults_table)
 If you want to use all available linters, you can include this in your `.lintr` file:
 
 ```r
 linters: linters_with_defaults(
-    defaults = linters_with_tags(tags = NULL)
+    defaults = all_linters()
   )
 ```
````
this should also directly call `all_linters()`.
Is there a section mentioning `linters_with_tags()`?

IndrajeetPatil commented:
I will fix the vignette later today.

AshesITR reviewed (approved):
Nice work, thanks!

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -194,29 +195,49 @@ defaults_table <- data.frame(
+
+You can include tag-based linters in the configuration file, and customize them further:
+
+```r
````
I just noticed, these config examples should be ```` ``` yaml ```` for consistency with ll. 77 and clearer separation of R code examples and config examples.

IndrajeetPatil replied:
Good idea. Done.

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -194,29 +195,49 @@ defaults_table <- data.frame(
+The available tags are listed below:
+
+```{r}
+available_tags(packages = "lintr")
````
lintr::

AshesITR commented on vignettes/lintr.Rmd:
````diff
@@ -194,29 +195,49 @@ defaults_table <- data.frame(
+You can select tags of interest to see which linters are included:
+
+```{r}
+linters <- linters_with_tags(tags = c("package_development", "readability"))
````
lintr::
